### ISSUE TITLE
Consider improvements to the op-builder auto-complete

### ISSUE THREAD

**Issue by hinerm** [scijava/scijava#111]
Auto-complete in an IDE is a very useful way to explore a new API. My understanding is that this is part of the intent of the `OpBuilder`: exposing methods incrementally to guide the user to completion of an Op call.

However, this is what the `OpBuilder` auto-complete menu looks like in practice:
![image](https://user-images.githubusercontent.com/1123809/213754288-0e4e0701-c2dd-4ec4-a356-dd44f2abfa47.png)
![image](https://user-images.githubusercontent.com/1123809/213754360-c4c81ecb-743b-4196-a382-309b387a6f76.png)
![image](https://user-images.githubusercontent.com/1123809/213754424-3e3a31a9-718a-4978-8eb0-62c851e50fd2.png)

This is overwhelming and distracting to me. Some thoughts:

* The acronyms, `IT`/`OU`/`OT`/etc... could just be written out verbosely. I can appreciate trying to save space, but when you have 60+ items in a menu it's more important to me to be able to distinguish them meaningfully.
* Fundamentally I have two options at this step: `input` or `inType` with a variable number of arguments. All I really want is just one of each of these to be at the top of my suggestions list.
* Because the `input` from `Arity0` is at the top of the list, if I just hit `enter` my cursor will skip to the end of the parens and I won't be able to type what my actual inputs are without backtracking. I'm going to wager that more often than not we are running Ops on *some* sort of inputs. I would much prefer my cursor to end inside the parens and then be able to start typing my inputs.
* I can understand the appeal of displaying all the different arities, but really I'm used to just typing params in a comma-separated way. This list of arities isn't actually tied to the Op I called so there's no correlation if most of these options make sense (and in fact, I believe most will **not** be applicable for a given Op).

So, I would personally prefer to just see a single-param or varargs `input` and `inType` at the top of the suggestions, potentially followed by zero-param versions. I think that would be both sufficient and more useful.

**Comment by ctrueden** [comment]
> I would personally prefer to just see a single-param or varargs input and inType at the top of the suggestions, potentially followed by zero-param versions. I think that would be both sufficient and more useful.

We can't use varargs because then we would lose the type safety the builder provides. A huge advantage of the builder is that you get back the exact generic type of the op you built—it's why the builder was invented in the first place. I do not know how to accomplish this without a large number of autocomplete options.

Like I said a few days ago, we could try to reduce the number of included functional interface arities to something smaller than 16-arity, but IIRC there was a good reason that @gselzer needed to boost up that high: because we had actual ops we were implementing with that many parameters.

It makes more sense to me to focus on adding type-unsafe varargs `ops.apply` and `ops.compute` directly to the `OpEnvironment`.

**Comment by ctrueden** [comment]
Here is one idea that could reduce the number of initial autocomplete options.

We could make:
```java
ops.op2("math.add").<TAB>
```
yield only binary autocompletes.

The downside would be that `ops.op<TAB>` would then have at least 17 autocompletes numbered 0 through 16 or 17.

What do you think @hinerm, do you think that would be more or less confusing than what we have now?

**Comment by gselzer** [comment]
> Like I said a few days ago, we could try to reduce the number of included functional interface arities to something smaller than 16-arity, but IIRC there was a good reason that @gselzer needed to boost up that high: because we had actual ops we were implementing with that many parameters.

Yeah, we had to deal with [this behemoth](https://github.com/scijava/incubator/blob/ad1535415a4dd36eb5837ccde9e6e7bca749e42f/imagej/imagej-ops2/src/main/java/net/imagej/ops2/deconvolve/RichardsonLucyC.java#L61), I think I just decided to go to the next highest power of 2 for fun.

> It makes more sense to me to focus on adding type-unsafe varargs ops.apply and ops.compute directly to the OpEnvironment.

Sure. We already go most of the way with e.g. [`FunctionUtils`](https://github.com/scijava/incubator/blob/ad1535415a4dd36eb5837ccde9e6e7bca749e42f/scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/util/FunctionUtils.java#L32) (which I forgot existed), it wouldn't be hard to wrap that within the `OpEnvironment`.

**Comment by gselzer** [comment]
> 

I personally think this is confusing/distracting. Maybe it's just the example you chose, but `op2` makes me think of versioning. But I also don't envision autocomplete being very useful for Ops, so maybe I'm not the one to judge.

**Comment by ctrueden** [comment]
@gselzer How about:
```
ops.unary("math.add").<TAB>
ops.binary("math.add").<TAB>
ops.arity3("math.add").<TAB>
ops.arity4("math.add").<TAB>
```
etc.?

**Comment by gselzer** [comment]
> @gselzer How about:
> ```java
> ops.unary("math.add").<TAB>
> ops.binary("math.add").<TAB>
> ops.arity3("math.add").<TAB>
> ops.arity4("math.add").<TAB>
> ```
> etc.?

Much better. I dislike the switch from unary, binary to arity3,4,... but it may be the best of the options

**Comment by ctrueden** [comment]
> dislike the switch from unary, binary to arity3,4

We could include both, with unary being an alias for arity1, etc.

**Comment by hinerm** [comment]
> ```java
> ops.unary("math.add").<TAB>
> ops.binary("math.add").<TAB>
> ops.arity3("math.add").<TAB>
> ops.arity4("math.add").<TAB>
> ```

I have some concerns about this. I am not a fan of the `arity_` naming, in general - when things are only differentiated by a number they can be hard to parse. It is tolerable when it's hidden in the interface names, but again, when I see a dialog like this, it does not feel like it's particularly helpful in guiding me to a choice:
![image](https://user-images.githubusercontent.com/1123809/214864676-06b9bbac-aa45-49ed-91d5-dda1f87b5b40.png)

Practically, if the `op` method was overloaded to parameter-specific naming, I anticipate being annoyed that I have to delete the whole method call if I want to change a parameter. I already dislike having different `apply` vs. `compute` methods. I appreciate the concise nature of a complete op call, but I feel like I need significant outside knowledge about the inner workings of the ops framework to make decisions in building that call.

> We can't use varargs because then we would lose the type safety the builder provides. A huge advantage of the builder is that you get back the exact generic type of the op you built—it's why the builder was invented in the first place. I do not know how to accomplish this without a large number of autocomplete options.

Maybe I just have a mismatch of expectations here. Maybe we should just make an `ops.run` type unsafe option as the recommended entry point for new users and anyone who doesn't need type safety. Then optimizing the accessibility of the builder would feel less important to me.

**Comment by gselzer** [comment]
See also #29 for related discussion

**Comment by ctrueden** [comment]
> Maybe I just have a mismatch of expectations here.

I think so, yeah. The builder exists to facilitate type safety—i.e. precision specification of the behavior you need in your program. You need a `math.add` that accepts two `Img<UnsignedByteType>` objects as arguments and writes the result into an `Img<FloatType>`? You can ask the builder for a binary computer op that does that, with the correct generic typing and correct functional interface. The builder syntax we have now was the nicest way we could conceive of to meet that precision specification requirement. It is also an important design goal of Ops v2 to use dedicated functional interfaces that can be invoked directly instead of some super-general-but-type-unsafe `Object run(Object... args)` and/or reflection-based approach, because the former's type safety also yields much better performance with the JIT.

I believe it is unlikely you will get the level of simplicity you want in Java from an IDE, without dedicated IDE plugins. Even if we hand-wave away the issue being discussed here, there is then also the fact that you cannot, using auto-complete, see which actual op implementations are available to autocomplete those types; you just have to ask and cross your fingers that what you need will be present at runtime. And even with a dedicated IDE plugin, there is yet another issue with the simplification/adaptation/et. al steps where a request you make could actually work but not be visible as a valid option in the autocommplete—there are an exponential number of such options 

> Maybe we should just make an ops.run type unsafe option as the recommended entry point for new users and anyone who doesn't need type safety.

We probably can't make `ops.run` specifically, as discussed in #29. We can make `ops.apply` and `ops.compute`. I think having only those two would be sufficient to always differentiate between function/computer/inplace1/inplace2/etc. in a nice way. Programmer using ops do need to learn the difference between function/apply and computer/compute—it's a critical distinction. Or I guess we could have only `compute(out, args...)` similar to what Ops v1 had, and use `apply` under the hood when `out=null`. I seem to recall people complaining about `out=null` being confusing previously, but maybe it would be OK.

**Comment by hinerm** [comment]
> See also #29 for related discussion

The highlights of this discussion, to me:

>I could imagine that a) burdens the user too much with implementation specifics (i.e., they need to know whether a given Op is a function, computer, or inplace) and makes running Ops less convenient.

⬆️ this is an essential point. I believe non-programmers need a path to running Ops that does not expose them to functions/computers/inplaces.

>c = a + b (function), or a += b (inplace), or "fill preallocated c data structure with the result of a + b" (computer). Writing ops.run("math.add", c, a, b) is semantically ambiguous

I think at this point we have established the requirement that modifiable parameters come last so the call would be `ops.run("math.add", a, b, c)` - and I think we can say the *intent* of this call is unambiguous from the user's point of view: to call `math.add` on `a` and `b` and store the result in `c`. Couldn't `ops.run` do successive match calls, e.g. first as `function` and then `computer` if the first fails? I think we could either add iterative inplace searching, or just say "use the builder if you want to do inplaces." If `ops.run` matched a computer it could still return the preallocated output as a return value.

>I seem to recall people complaining about out=null being confusing previously, but maybe it would be OK.

I don't think it's ok. Every piece of added information that is required is another barrier to use, and I think we should have a path to use of Ops that aggressively removes every such barrier that we can.

>Programmer using ops do need to learn the difference between function/apply and computer/compute—it's a critical distinction.

I can *maybe* accept that requirement for **programmers**. If I was a non-technical user and was told I had to understand these concepts to use Ops, I would probably just use the IJ1 macro recorder.

For programmers, I maintain that the points on my first post all make writing ops calls with the builder annoying. Do we have a chart of the decision points of the builder? I think that might be helpful to discuss how to optimize the implementation.

**Comment by ctrueden** [comment]
> `ops.run("math.add", a, b, c)` - and I think we can say the *intent* of this call is unambiguous from the user's point of view: to call `math.add` on `a` and `b` and store the result in `c`.

No, they might mean "a + b + c". And my strong intuition is that there will be many such ambiguities across the system if we start trying to do that sort of iterative matching. E.g. places where the person wants a unary computer but there also exists a matching binary function which gets used instead, surprising the user.

In general, I don't understand this argument that the Java API needs to cater to non-technical users. If you are a non-technical user, do not program in Java. Use the GUI, or if you are a scripter use something like Python, which as previously discussed we can make simpler and more elegant.

**Comment by ctrueden** [comment]
> This list of arities isn't actually tied to the Op I called so there's no correlation if most of these options make sense (and in fact, I believe most will not be applicable for a given Op).

☝️ This point raises an interesting question about the design of Ops. Right now, a given op *name* such as `math.add` does not have any arity tied to it. But should it?

**Comment by gselzer** [comment]
> In general, I don't understand this argument that the Java API needs to cater to non-technical users. If you are a non-technical user, do not program in Java. Use the GUI, or if you are a scripter use something like Python, which as previously discussed we can make simpler and more elegant.

I agree. I think one major benefit to working in Java is type safety, and we should leverage that. There are arguably better libraries to work in (i.e. Python) if you don't care about type safety.



> point_up This point raises an interesting question about the design of Ops. Right now, a given op _name_ such as `math.add` does not have any arity tied to it. But should it?

I would argue strongly against that. Consider optional parameters. If you have a `Functions.Arity3` with an optional parameter, what would you call it?

**Comment by hinerm** [comment]
> ```java
> ops.unary("math.add").<TAB>
> ops.binary("math.add").<TAB>
> ops.arity3("math.add").<TAB>
> ops.arity4("math.add").<TAB>
> ```

If we are clear that the builder is for technical users/java developers and provide a different type-unsafe path to execution for non-technical users, I am more fond of something like this.

One thought is there could also be another intermediate step in the builder to differentiate *running* the Op vs. *returning* an Op, which would split the `input`/`inType` options so they don't pollute each other.

**Comment by hinerm** [comment]
See also #35

**Comment by gselzer** [comment]
@hinerm can this be closed?
